rust/main/chains/hyperlane-ethereum/src/rpc_clients/fallback/tests.rs

```diff
@@ -0,0 +1,216 @@
+    serde_json::from_str("not-a-json").map_err(|e| HttpClientError::SerdeJson {
+        err: e,
+        text: "".to_owned(),
+    })
+}
+
+#[async_trait]
+impl JsonRpcClient for EthereumProviderMock {
+    type Error = HttpClientError;
+
+    /// Pushes the `(method, params)` to the back of the `requests` queue,
+    /// pops the responses from the back of the `responses` queue
+    async fn request<T: Debug + Serialize + Send + Sync, R: DeserializeOwned>(
+        &self,
+        method: &str,
+        params: T,
+    ) -> Result<R, Self::Error> {
+        if let Some(sleep_duration) = self.provider.request_sleep() {
+            sleep(sleep_duration).await;
+        }
+        self.push(method, params);
+        if self.success {
+            dummy_success_return_value()
+        } else {
+            dummy_error_return_value()
+        }
+    }
+}
+
+impl PrometheusJsonRpcClientConfigExt for EthereumProviderMock {
+    fn node_host(&self) -> &str {
+        todo!()
+    }
+
+    fn chain_name(&self) -> &str {
+        todo!()
+    }
+}
+
+impl<C> EthereumFallbackProvider<C, JsonRpcBlockGetter<C>>
+where
+    C: JsonRpcClient<Error = HttpClientError>
+        + PrometheusJsonRpcClientConfigExt
+        + Into<JsonRpcBlockGetter<C>>
+        + Clone,
+    JsonRpcBlockGetter<C>: BlockNumberGetter,
+{
+    async fn fallback_test_call(&self) {
+        self.request::<_, u64>(BLOCK_NUMBER_RPC, ()).await.unwrap();
+    }
+}
+
+#[tokio::test]
+async fn test_multicast_first_provider_succeeds() {
+    let fallback_provider_builder = FallbackProviderBuilder::default();
+    let providers = vec![
+        EthereumProviderMock::default(),
+        EthereumProviderMock::default(),
+        EthereumProviderMock::default(),
+    ];
+    let fallback_provider = fallback_provider_builder.add_providers(providers).build();
+    let ethereum_fallback_provider = EthereumFallbackProvider::new(fallback_provider);
+    ethereum_fallback_provider
+        .multicast::<_, u64>(BLOCK_NUMBER_RPC, ())
+        .await
+        .unwrap();
+    let provider_call_count: Vec<_> =
+        ProviderMock::get_call_counts(&ethereum_fallback_provider).await;
+    assert_eq!(provider_call_count, vec![1, 0, 0]);
+}
+
+#[tokio::test]
+async fn test_multicast_second_provider_succeeds() {
+    let fallback_provider_builder = FallbackProviderBuilder::default();
+    let providers = vec![
+        EthereumProviderMock::new(None, false),
+        EthereumProviderMock::default(),
+        EthereumProviderMock::default(),
+    ];
+    let fallback_provider = fallback_provider_builder.add_providers(providers).build();
+    let ethereum_fallback_provider = EthereumFallbackProvider::new(fallback_provider);
+    ethereum_fallback_provider
+        .multicast::<_, u64>(BLOCK_NUMBER_RPC, ())
+        .await
+        .unwrap();
+    let provider_call_count: Vec<_> =
+        ProviderMock::get_call_counts(&ethereum_fallback_provider).await;
+    assert_eq!(provider_call_count, vec![1, 1, 0]);
```

Comment:
ditto do we guarantee these first 2 if we are using `FuturesUnordered`?

feels like these test cases aren't very meaningfully different from the behavior we'd expect with a fallback, is there any other way to test that all providers are requested concurrently?

Reply:
It is mostly the implementation detail of `FutureUnordered`.

I'll change the test case to simulate that we call all mock before anyone succeeds.